cpp/include/raft/neighbors/cagra_types.hpp

```diff
@@ -258,6 +258,13 @@ struct index : ann::index {
         dataset.data_handle(), dataset.extent(0), dataset.extent(1), dataset.extent(1));
     }
   }
+  void update_dataset(raft::resources const& res,
```

Comment:
We probably want to keep these const mdspans. If this is because of python, can we use make_const_mdspan() in that layer?

Reply:
I agree that automatically discarding `const` would be bad - but this is doing the opposite and is automatically adding it (like this is converting a non-const mdspan to a const msdpan), which I feel like is something that should be allowed with our API's. 

The issue I have is that Cython kinda sucks with respecting `const` identifiers, which is why all our Cython api's use non-const mdspans right now. Like if I try to add a `get_const_hmv_float` (to parallel the non-const `get_hmv_float` we have now) - I get an error message from Cython, where it doesn't recognize `const float` as a type inside template parameters:

```
      Error compiling Cython file:
      ------------------------------------------------------------
      ...
          if cai.dtype != np.float32:
              raise TypeError("dtype %s not supported" % cai.dtype)
          if check_shape and len(cai.shape) != 2:
              raise ValueError("Expected a 2D array, got %d D" % len(cai.shape))
          shape = (cai.shape[0], cai.shape[1] if len(cai.shape) == 2 else 1)
          return make_host_matrix_view[const float, int64_t, row_major](
                                             ^
      ------------------------------------------------------------
      
      /home/ben/code/raft/python/pylibraft/pylibraft/common/mdspan.pyx:232:39: Expected ']', found 'float'
```

I can get around this by adding a Cython typedef (like `ctypedef const float const_float`) - but that introduces the need for other hacks later on (like cython will treat `const_float` and `const float` as separate types - meaning that when we define the `update_dataset` for Cython in c_cagra.pxd I can't just go `const T` as the type, and have to introduce a new template param =(.  I've done this in the last commit - let me know what you think